cpp/test/preprocess_utils.cu

```diff
@@ -0,0 +1,283 @@
+      float val = host_matrix(row, col);
+      if (val == 0) {
+        out_host_matrix(row, col) = 0.0f;
+      } else {
+        float tf  = float(val / h_output_cols_lengths(0, col));
+        float idf = raft::log<float>(num_cols / h_output_rows_cnt(0, row));
+        if (tf_idf) {
+          result = tf * idf;
+        } else {
+          float bm25 = ((k1 + 1) * tf) /
+                       (k1 * ((1 - b) + b * (h_output_cols_lengths(0, col) / avg_col_length)) + tf);
+          result = idf * bm25;
+        }
+        out_host_matrix(row, col) = result;
+        out_host_vector(count)    = result;
+        count++;
+      }
+    }
+  }
+  raft::copy(results.data_handle(), out_host_vector.data_handle(), out_host_vector.size(), stream);
+}
+
+template <typename T1, typename T2>
+int get_dupe_mask_count(raft::resources& handle,
+                        raft::device_vector_view<T1> rows,
+                        raft::device_vector_view<T1> columns,
+                        raft::device_vector_view<T2> values,
+                        const raft::device_vector_view<T1>& mask)
+{
+  cudaStream_t stream = raft::resource::get_cuda_stream(handle);
+
+  raft::sparse::op::coo_sort(rows.size(),
+                             columns.size(),
+                             values.size(),
+                             rows.data_handle(),
+                             columns.data_handle(),
+                             values.data_handle(),
+                             stream);
+
+  raft::sparse::op::compute_duplicates_mask<int>(
+    mask.data_handle(), rows.data_handle(), columns.data_handle(), rows.size(), stream);
+
+  int col_nnz_count = thrust::reduce(raft::resource::get_thrust_policy(handle),
+                                     mask.data_handle(),
+                                     mask.data_handle() + mask.size());
+  return col_nnz_count;
+}
+
+template <typename T1, typename T2>
+void remove_dupes(raft::resources& handle,
+                  raft::device_vector_view<T1> rows,
+                  raft::device_vector_view<T1> columns,
+                  raft::device_vector_view<T2> values,
+                  raft::device_vector_view<T1> mask,
+                  const raft::device_vector_view<T1>& out_rows,
+                  const raft::device_vector_view<T1>& out_cols,
+                  const raft::device_vector_view<T2>& out_vals,
+                  int num_rows = 128)
+{
+  cudaStream_t stream = raft::resource::get_cuda_stream(handle);
+
+  auto col_counts = raft::make_device_vector<int, int64_t>(handle, columns.size());
+
+  thrust::fill(raft::resource::get_thrust_policy(handle),
+               col_counts.data_handle(),
+               col_counts.data_handle() + col_counts.size(),
+               1.0f);
+
+  auto keys_out   = raft::make_device_vector<int, int64_t>(handle, num_rows);
+  auto counts_out = raft::make_device_vector<int, int64_t>(handle, num_rows);
+
+  thrust::reduce_by_key(raft::resource::get_thrust_policy(handle),
```

Comment:
We have a great function already for removing duplicates from sparse formats- it uses a simple mask to figure out where the duplicates are. It's really efficient. Also, if the goal is to get the degree for each row of the matrix, we have functions for this.

Reply:
I'm really less concerned about thrust in tests... however it does make it easier if we can reuse raft routines

Reply:
So the two available functions for removing duplicates that I saw are `compute_duplicates_mask` and `max_duplicates` now I did not find one that takes a mask and removes based on it. And what I have here is a function that uses the mask to remove the dupes. Max duplicates works a little different than the mask. It will opt to leave the max value row, however that is not the behavior I want. I would like to take the last vertice value that we see in the COO vectors. This aligns more with the compute_duplicates_mask function which is what I used here. But all this other stuff is required. If you look further up in the code you will see that before we remove the dupes we use that function to calculate the mask. The mask is actually used in this function. If there is a function that exists already that takes this mask directly and will remove the 0 value indices, I would love to use it. I could not find it though.